XiEditor/Trace.swift

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A simple mechanism for logging trace events and outputting a file in
+// Chrome tracing format
+
+import Foundation
+
+var globalTrace = Trace()
+
+class Trace {
+    let mutex = UnfairLock()
+    let BUF_SIZE = 100_000
+    var buf: [TraceEntry]
+    var n_entries = 0
+
+    init() {
+        buf = [TraceEntry](repeating: TraceEntry(), count: BUF_SIZE)
```

Comment:
Musing: This could be set at declare time if we're willing to make it lazy, or if we make BUF_SIZE a type property; then we wouldn't need the init. (Although we might want one for other reasons... ) 🤷‍♂️